File: serenityboard/server/run_manager.py
"""Run discovery and lifecycle management via filesystem polling."""

from __future__ import annotations

import asyncio
import logging
import os
import shutil
import time

from serenityboard.server.data_provider import RunDataProvider

# If a "running" run has no new data for this many seconds, mark it "completed".
STALE_TIMEOUT_S = 300  # 5 minutes

__all__ = ["RunWatcher"]

logger = logging.getLogger(__name__)


class RunWatcher:
    """Periodically scans logdir for new board.db files. Detects new/removed runs.

    Uses filesystem polling (5s default). Each subdirectory of *logdir* that
    contains a ``board.db`` file is treated as a run. The subdirectory name
    becomes the run name.
    """

    def __init__(self, logdir: str, poll_interval: float = 5.0) -> None:
        self.logdir = logdir
        self.poll_interval = poll_interval
        self.known_runs: dict[str, RunDataProvider] = {}  # run_name -> provider
        self._callbacks_add: list = []  # callbacks for new runs
        self._callbacks_remove: list = []  # callbacks for removed runs

    # ------------------------------------------------------------------
    # Callback registration
    # ------------------------------------------------------------------

    def on_add(self, callback) -> None:
        """Register callback for new run detection."""
        self._callbacks_add.append(callback)

    def on_remove(self, callback) -> None:
        """Register callback for run removal."""
        self._callbacks_remove.append(callback)

    # ------------------------------------------------------------------
    # Scanning
    # ------------------------------------------------------------------

    def _find_run_dbs(self) -> dict[str, str]:
        """Recursively find board.db files up to 4 levels deep.

        Returns {run_name: db_path} where run_name is a URL-safe key
        derived from the relative path (slashes replaced with ``__``).
        Falls back to basename when the db sits directly inside logdir.
        """
        found: dict[str, str] = {}
        if not os.path.isdir(self.logdir):
            return found
        logdir = self.logdir.rstrip(os.sep) + os.sep
        for root, dirs, files in os.walk(self.logdir):
            # Limit depth to 4 levels
            depth = root[len(self.logdir):].count(os.sep)
            if depth >= 4:
                dirs.clear()
                continue
            if "board.db" in files:
                db_path = os.path.join(root, "board.db")
                rel = root[len(logdir):]
                if rel:
                    # Replace path separators so run name is URL-safe.
                    run_name = rel.replace(os.sep, "__")
                else:
                    run_name = os.path.basename(root)
                found[run_name] = db_path
        return found

    def scan_once(self) -> tuple[list[str], list[str]]:
        """Synchronous scan. Returns (added, removed) run names."""
        added: list[str] = []
        removed: list[str] = []

        found = self._find_run_dbs()
        current = set(found.keys())

        for name, db_path in found.items():
            if name not in self.known_runs:
                try:
                    provider = RunDataProvider(db_path)
                    provider.get_run_info()
                    self.known_runs[name] = provider
                    added.append(name)
                except Exception:
                    logger.warning("Failed to open %s", db_path)

        gone = set(self.known_runs) - current
        for name in gone:
            provider = self.known_runs.pop(name)
            provider.close()
            removed.append(name)

        return added, removed

    async def scan_loop(self) -> None:
        """Background async scan loop."""
        while True:
            added, removed = self.scan_once()
            for name in added:
                for cb in self._callbacks_add:
                    if asyncio.iscoroutinefunction(cb):
                        await cb(name)
                    else:
                        cb(name)
            for name in removed:
                for cb in self._callbacks_remove:
                    if asyncio.iscoroutinefunction(cb):
                        await cb(name)
                    else:
                        cb(name)
            await asyncio.sleep(self.poll_interval)

    # ------------------------------------------------------------------
    # Query helpers
    # ------------------------------------------------------------------

    def get_runs(self) -> list[dict]:
        """Return list of run info dicts for all known runs.

        Auto-detects stale runs and classifies as:
        - "completed": reached max_steps (or writer.close() was called)
        - "stopped":   no new data for 5min but didn't reach max_steps
        - "running":   actively logging
        """
        now = time.time()
        result: list[dict] = []
        for name, provider in sorted(self.known_runs.items()):
            try:
                info = provider.get_run_info()
                status = info.get("status", "unknown")
                last_wt, last_step = provider.get_last_activity()
                hparams = info.get("hparams", {})
                max_steps = hparams.get("max_steps")

                # Auto-detect finished/stopped runs that never called writer.close()
                if status == "running":
                    if last_wt is not None and now - last_wt > STALE_TIMEOUT_S:
                        if max_steps and last_step is not None and last_step >= max_steps:
                            status = "completed"
                        else:
                            status = "stopped"
                    elif last_wt is None:
                        # Never logged any data — check if start_time is stale
                        st = info.get("start_time")
                        if st is not None and now - st > STALE_TIMEOUT_S:
                            status = "empty"

                result.append({
                    "name": name,
                    "start_time": info.get("start_time"),
                    "status": status,
                    "last_activity": last_wt,
                    "last_step": last_step,
                    "max_steps": max_steps,
                    "active_session_id": info.get("active_session_id"),
                    "hparams": hparams,
                })
            except Exception:
                result.append({
                    "name": name,
                    "start_time": None,
                    "status": "error",
                    "last_activity": None,
                    "last_step": None,
                    "max_steps": None,
                    "active_session_id": None,
                    "hparams": {},
                })
        return result

    def get_provider(self, run_name: str) -> RunDataProvider | None:
        """Return the data provider for a specific run, or None."""
        return self.known_runs.get(run_name)

    def delete_run(self, run_name: str) -> bool:
        """Delete a run's data directory and remove from registry.

        Returns True if deleted, False if not found.
        """
        provider = self.known_runs.get(run_name)
        if provider is None:
            return False
        run_dir = provider.run_dir
        provider.close()
        del self.known_runs[run_name]
        if os.path.isdir(run_dir):
            shutil.rmtree(run_dir)
        return True

    # ------------------------------------------------------------------
    # Cleanup
    # ------------------------------------------------------------------

    def close(self) -> None:
        """Close all providers and clear state."""
        for provider in self.known_runs.values():
            provider.close()
        self.known_runs.clear()

File: native/src/server/run_manager.cpp
#include "serenityboard/run_manager.hpp"

#include <filesystem>
#include <functional>

#include "serenityboard/session_guard.hpp"

namespace sb {

namespace fs = std::filesystem;

std::map<std::string, std::string> RunWatcher::find_run_dbs() const {
  std::map<std::string, std::string> found;
  std::error_code ec;
  if (!fs::is_directory(logdir_, ec)) return found;
  std::string base = logdir_;
  while (base.size() > 1 && base.back() == '/') base.pop_back();
  // os.walk with depth limit: depth = root[len(logdir):].count('/'); stop descending at depth >= 4
  std::function<void(const fs::path &, int)> walk = [&](const fs::path &root, int depth) {
    if (depth >= 4) return;
    std::error_code e;
    if (fs::exists(root / "board.db", e)) {
      const std::string root_text = root.string();
      std::string rel = root_text.size() > base.size() + 1 ? root_text.substr(base.size() + 1) : "";
      std::string name;
      if (!rel.empty()) {
        name = rel;
        std::size_t pos;
        while ((pos = name.find('/')) != std::string::npos) name.replace(pos, 1, "__");
      } else {
        name = root.filename().string();
      }
      found[name] = (root / "board.db").string();
    }
    for (const auto &entry : fs::directory_iterator(root, e)) {
      if (entry.is_directory(e)) walk(entry.path(), depth + 1);
    }
  };
  walk(fs::path(base), 0);
  return found;
}

std::pair<std::vector<std::string>, std::vector<std::string>> RunWatcher::scan_once() {
  std::vector<std::string> added, removed;
  const auto found = find_run_dbs();
  std::lock_guard<std::mutex> lock(mutex_);
  for (const auto &[name, db_path] : found) {
    if (known_runs_.count(name)) continue;
    try {
      auto provider = std::make_shared<RunDataProvider>(db_path);
      provider->get_run_info();
      known_runs_[name] = provider;
      added.push_back(name);
    } catch (const std::exception &) {
      // Failed to open (mid-creation, corrupt): retry on the next scan.
    }
  }
  for (auto it = known_runs_.begin(); it != known_runs_.end();) {
    if (!found.count(it->first)) {
      it->second->close();
      removed.push_back(it->first);
      it = known_runs_.erase(it);
    } else {
      ++it;
    }
  }
  return {added, removed};
}

Json RunWatcher::get_runs() {
  const double now = wall_time_now();
  std::vector<std::pair<std::string, std::shared_ptr<RunDataProvider>>> snapshot;
  {
    std::lock_guard<std::mutex> lock(mutex_);
    for (const auto &[name, p] : known_runs_) snapshot.emplace_back(name, p);
  }
  Json result = Json::array();
  for (const auto &[name, provider] : snapshot) {
    try {
      Json info = provider->get_run_info();
      std::string status = info.contains("status") && info["status"].is_string() ? info["status"].get<std::string>() : "unknown";
      const auto [last_wt, last_step] = provider->get_last_activity();
      Json hparams = info.contains("hparams") && info["hparams"].is_object() ? info["hparams"] : Json::object();
      Json max_steps = hparams.contains("max_steps") ? hparams["max_steps"] : Json(nullptr);
      if (status == "running") {
        if (last_wt && now - *last_wt > kStaleTimeoutSeconds) {
          const bool reached = max_steps.is_number() && last_step && static_cast<double>(*last_step) >= max_steps.get<double>();
          status = reached ? "completed" : "stopped";
        } else if (!last_wt) {
          if (info.contains("start_time") && info["start_time"].is_number() && now - info["start_time"].get<double>() > kStaleTimeoutSeconds)
            status = "empty";
        }
      }
      result.push_back({{"name", name},
                        {"start_time", info.contains("start_time") ? info["start_time"] : Json(nullptr)},
                        {"status", status},
                        {"last_activity", last_wt ? Json(*last_wt) : Json(nullptr)},
                        {"last_step", last_step ? Json(*last_step) : Json(nullptr)},
                        {"max_steps", max_steps},
                        {"active_session_id", info.contains("active_session_id") ? info["active_session_id"] : Json(nullptr)},
                        {"hparams", hparams}});
    } catch (const std::exception &) {
      result.push_back({{"name", name}, {"start_time", nullptr}, {"status", "error"}, {"last_activity", nullptr},
                        {"last_step", nullptr}, {"max_steps", nullptr}, {"active_session_id", nullptr},
                        {"hparams", Json::object()}});
    }
  }
  return result;
}

std::shared_ptr<RunDataProvider> RunWatcher::get_provider(const std::string &run_name) {
  std::lock_guard<std::mutex> lock(mutex_);
  auto it = known_runs_.find(run_name);
  return it == known_runs_.end() ? nullptr : it->second;
}

bool RunWatcher::delete_run(const std::string &run_name) {
  std::shared_ptr<RunDataProvider> provider;
  {
    std::lock_guard<std::mutex> lock(mutex_);
    auto it = known_runs_.find(run_name);
    if (it == known_runs_.end()) return false;
    provider = it->second;
    known_runs_.erase(it);
  }
  const std::string run_dir = provider->run_dir();
  provider->close();
  std::error_code ec;
  if (fs::is_directory(run_dir, ec)) fs::remove_all(run_dir, ec);
  return true;
}

void RunWatcher::close() {
  std::lock_guard<std::mutex> lock(mutex_);
  for (auto &[_, p] : known_runs_) p->close();
  known_runs_.clear();
}

}  // namespace sb
